### PR TITLE
[FLOC-3692] Failing test for protecting the dataset paths.

### PR DESCRIPTION
**Pull request by sarum90**
Cleaned up version of #2339

Note that this branches off of make-test-simple-transitions-work-FLOC-3713 because it uses the test infrastructure that was fixed in that branch.

https://clusterhq.atlassian.net/browse/FLOC-3692

This is a failing test for FLOC-3254.

That issue is to protect against docker creating and using a mountpath for a dataset when flocker hasn't actually manifested the dataset. The method for testing is to keep trying to use the path reported by flocker as soon as it is reported, and verify that there is never a time where we have a successful write to that path that is not reflected in the backing blockdevice. We check this repeatedly before and after any use of the BlockdeviceManager interface.

Note that I'm not merging against master instead choosing to merge against a different branch specific to this issue.

I'd like to use this branch to gain confidence in the approach to the fix prior to having to tackle all of the details involved in upgrading flocker from where we currently mount datasets to where we will mount datasets after fixing FLOC-3254.

### CODE REVIEW

File: flocker/node/agents/test/test_blockdevice.py [hunk 1]
```diff
@@ -1195,29 +1203,156 @@ class DidNotConverge(Exception):
...
+    def invariant(self, case, current_cluster_state):
+        """
+        An invariant that should always hold. Specifically, after the path
+        to a dataset is known by the cluster then writes to that path
+        should either succeed and be persisted to the underlying block
+        device, or the write should fail.
+
+        :param case: A ``TestCase`` to use for making assertions.
+        :param current_cluster_state: A callable that returns the current state
+            of the cluster.
```
Comment (mpgomez) [inline flocker/node/agents/test/test_blockdevice.py]: return?

File: flocker/node/agents/test/test_blockdevice.py [hunk 2]
```diff
@@ -1195,29 +1203,156 @@ class DidNotConverge(Exception):
...
+        """
+        thing = False
```
Comment (mpgomez) [inline flocker/node/agents/test/test_blockdevice.py]: It is a big difficult to understand what "thing" is

Comment (sarum90) [inline flocker/node/agents/test/test_blockdevice.py]: Thanks, I didn't mean to have that variable anymore.

File: flocker/node/agents/test/test_blockdevice.py (hunk 1 above)
Comment (sarum90) [inline flocker/node/agents/test/test_blockdevice.py]: Supposed to return nothing :) -- the return was left over from some debugging code.

**Comment by sarum90** [comment]
I sort of want this to be merged to non-master. Closing and re-openning.

**Pull request closed without merging**